fuse_models/src/april_tag_pose.cpp

```diff
@@ -0,0 +1,150 @@
+{
+  interfaces_ = interfaces;
+  fuse_core::AsyncSensorModel::initialize(interfaces, name, transaction_callback);
+}
+
+void AprilTagPose::onInit()
+{
+  logger_ = interfaces_.get_node_logging_interface()->get_logger();
+  clock_ = interfaces_.get_node_clock_interface()->get_clock();
+
+  // Read settings from the parameter sever
+  device_id_ = fuse_variables::loadDeviceId(interfaces_);
+
+  params_.loadFromROS(interfaces_, name_);
+
+  throttled_callback_.setThrottlePeriod(params_.throttle_period);
+
+  if (!params_.throttle_use_wall_time)
+  {
+    throttled_callback_.setClock(clock_);
+  }
+
+  if (params_.position_indices.empty() && params_.orientation_indices.empty())
+  {
+    RCLCPP_WARN_STREAM(logger_,
+                       "No dimensions were specified. Data from topic " << params_.topic << " will be ignored.");
+  }
+
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(clock_);
+  tf_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf_buffer_, interfaces_.get_node_base_interface(),
+                                                              interfaces_.get_node_logging_interface(),
+                                                              interfaces_.get_node_parameters_interface(),
+                                                              interfaces_.get_node_topics_interface());
+}
+
+void AprilTagPose::onStart()
+{
+  if (!params_.position_indices.empty() || !params_.orientation_indices.empty())
+  {
+    rclcpp::SubscriptionOptions sub_options;
+    sub_options.callback_group = cb_group_;
+
+    sub_ = rclcpp::create_subscription<MessageType>(interfaces_, params_.topic, params_.queue_size,
+                                                    std::bind(&AprilTagThrottledCallback::callback<const MessageType&>,
+                                                              &throttled_callback_, std::placeholders::_1),
+                                                    sub_options);
+  }
+}
+
+void AprilTagPose::onStop()
+{
+  sub_.reset();
+}
+
+void AprilTagPose::process(MessageType const& msg)
+{
+  for (auto const& transform : msg.transforms)
+  {
+    // Create a transaction object
+    auto transaction = fuse_core::Transaction::make_shared();
+    transaction->stamp(transform.header.stamp);
+
+    // Create the pose from the transform
+    auto pose = std::make_unique<geometry_msgs::msg::PoseWithCovarianceStamped>();
+    pose->header = transform.header;
+    pose->header.frame_id = pose->header.frame_id;
+    pose->pose.pose.orientation = transform.transform.rotation;
+    pose->pose.pose.position.x = transform.transform.translation.x;
+    pose->pose.pose.position.y = transform.transform.translation.y;
+    pose->pose.pose.position.z = transform.transform.translation.z;
+
+    // TODO(henrygerardmoore): figure out better method to set the covariance
+    for (std::size_t i = 0; i < params_.pose_covariance.size(); ++i)
+    {
+      pose->pose.covariance[i * 7] = params_.pose_covariance[i];
+    }
```

Comment:
IMO for_each is clunky here since we're doing the `i * 7` thing (to fill the diagonal elems of the covariance mat). I'm sure there's a way to do it, but I think using indices here is cleaner.